python-api/tests/integrationtests/test_controller.py

```diff
@@ -3,725 +3,343 @@
+        assert self.controller.get_encode_port() == 3002
 
-            sources = TestSources(video_port=3000)
-            sources.new_test_video()
-            sources.new_test_video()
 
-            GLib.timeout_add_seconds(5, self.quit_mainloop)
-            self.run_mainloop()
+class TestGetAudioPort(IntegrationTestbase):
+    """ Test get_audio_port method
+    """
+    def wait_until_ready(self):
+        """ Blocks until the Server has reported, that it's
+        audio-output has started
+        """
+        self.log.info("waiting for Server to start preview-port-outputs")
+        self.serv.wait_for_output('tcpserversink name=sink')
 
-            print(test_cb.call_args_list)
-            test_cb.assert_any_call(3003, 1, 7)
-            test_cb.assert_any_call(3004, 1, 8)
-            assert test_cb.call_count == 2
+    def test_audio_port(self):
+        """Test get_audio_port method returning the expected port"""
+        self.setup_server()
+        self.setup_controller()
 
-            serv.terminate(1)
-        finally:
-            serv.terminate_and_output_status(cov=True)
+        self.log.info("starting 1 test-audio source")
+        self.new_test_audio()
 
+        self.wait_until_ready()
 
-class VideoFileSink(object):
+        self.log.info("calling get_audio_port")
+        assert self.controller.get_audio_port() == 3003
 
-    """Sink the video to a file
+
+class TestGetPreviewPorts(IntegrationTestbase):
+    """ Test get_preview_ports method
     """
+    def wait_for_sources(self, count):
+        """ Blocks until the Server has reported, that the right number of
+        preview-ports are is started
+        """
+        self.log.info("waiting for Server to start preview-port-outputs")
+        self.serv.wait_for_output('tcpserversink name=sink', count=count)
+
+    def test_one_preview_port(self):
+        """Test get_preview_ports method returning the expected port"""
+        self.setup_server()
+        self.setup_controller()
+
+        self.log.info("starting 1 test-video source")
+        self.new_test_video()
+
+        self.wait_for_sources(count=1)
+
+        self.log.info("calling get_preview_ports")
+        assert self.controller.get_preview_ports() == [3003]
+
+    def test_five_preview_ports(self):
+        """Test get_preview_ports method returning the expected port"""
+        self.setup_server()
+        self.setup_controller()
+
+        self.log.info("starting 5 test-video sources")
+        self.new_test_video()
+        self.new_test_video()
+        self.new_test_video()
+        self.new_test_video()
+        self.new_test_video()
+
+        self.log.info("waiting for the test-video sources to come up")
+        self.wait_for_sources(count=5)
+
+        self.log.info("calling get_preview_ports")
+        assert self.controller.get_preview_ports() == [
+            3003, 3004, 3005, 3006, 3007]
+
+
+class TestSignals(IntegrationTestbaseMainloop):
```

Comment:
I haven't finished looking through anything yet but was wondering;

 * Do you have tests which handle the case where a controller is connected after everything it set up?
 * What about the case where multiple controllers are connected?